dll/win32/browseui/aclistisf.cpp

```diff
@@ -20,83 +21,411 @@
+                    return hr;
+                }
+            }
+            // FALL THROUGH
+        case LT_DESKTOP:
+            ++m_iNextLocation;
+            if (m_dwOptions & ACLO_DESKTOP)
+            {
+                CComHeapPtr<ITEMIDLIST> pidl;
+                hr = SHGetSpecialFolderLocation(NULL, CSIDL_DESKTOP, &pidl);
+                if (FAILED(hr))
+                    return S_FALSE;
+                hr = SetLocation(pidl);
+                if (SUCCEEDED(hr))
+                {
+                    TRACE("LT_DESKTOP\n");
+                    return hr;
+                }
+            }
+            // FALL THROUGH
+        case LT_MYCOMPUTER:
+            ++m_iNextLocation;
+            if (m_dwOptions & ACLO_MYCOMPUTER)
+            {
+                CComHeapPtr<ITEMIDLIST> pidl;
+                hr = SHGetSpecialFolderLocation(NULL, CSIDL_DRIVES, &pidl);
+                if (FAILED(hr))
+                    return S_FALSE;
+                hr = SetLocation(pidl);
+                if (SUCCEEDED(hr))
+                {
+                    TRACE("LT_MYCOMPUTER\n");
+                    return hr;
+                }
+            }
+            // FALL THROUGH
+        case LT_FAVORITES:
+            ++m_iNextLocation;
+            if (m_dwOptions & ACLO_FAVORITES)
+            {
+                CComHeapPtr<ITEMIDLIST> pidl;
+                hr = SHGetSpecialFolderLocation(NULL, CSIDL_FAVORITES, &pidl);
+                if (FAILED(hr))
+                    return S_FALSE;
+                hr = SetLocation(pidl);
+                if (SUCCEEDED(hr))
+                {
+                    TRACE("LT_FAVORITES\n");
+                    return hr;
+                }
+            }
+            // FALL THROUGH
+        case LT_MAX:
+        default:
+            TRACE("LT_MAX\n");
+            return S_FALSE;
+    }
+}
+
+HRESULT CACListISF::SetLocation(LPCITEMIDLIST pidl)
+{
+    TRACE("SetLocation(%p, %p)\n", this, pidl);
+
+    m_pEnumIDList.Release();
+    m_pShellFolder.Release();
+    m_pidlLocation.Free();
+
+    if (!pidl)
+        return E_FAIL;
+
+    LPITEMIDLIST pidlClone = ILClone(pidl);
+    if (!pidlClone)
+    {
+        ERR("Out of memory\n");
+        return E_OUTOFMEMORY;
+    }
+    m_pidlLocation.Attach(pidlClone);
+
+    CComPtr<IShellFolder> pFolder;
+    HRESULT hr = SHGetDesktopFolder(&pFolder);
+    if (FAILED(hr))
+    {
+        ERR("SHGetDesktopFolder failed: 0x%lX\n", hr);
+        return hr;
+    }
+
+    if (!ILIsEmpty(pidl))
+        pFolder->BindToObject(pidl, NULL, IID_PPV_ARG(IShellFolder, &m_pShellFolder));
+    else
+        m_pShellFolder.Attach(pFolder.Detach());
+
+    SHCONTF Flags = SHCONTF_FOLDERS | SHCONTF_INIT_ON_FIRST_NEXT;
+    if (m_fShowHidden)
+        Flags |= SHCONTF_INCLUDEHIDDEN;
+    if (!(m_dwOptions & ACLO_FILESYSDIRS))
+        Flags |= SHCONTF_NONFOLDERS;
+
+    hr = m_pShellFolder->EnumObjects(NULL, Flags, &m_pEnumIDList);
+    if (hr != S_OK)
```

Comment:
can this be S_FALSE?
Do you need to report that as E_FAIL?

Reply:
https://docs.microsoft.com/en-us/windows/win32/api/shobjidl_core/nf-shobjidl_core-ishellfolder-enumobjects

> Some implementations may also return S_FALSE, indicating that there are no children matching the grfFlags that were passed in. If S_FALSE is returned, ppenumIDList is set to NULL.

Reply:
`SetLocation` is used in `Reset`, `Expand`, and `NextLocation`. They expect `S_OK` or failure of `SetLocation`. `S_FALSE` of `EnumObjects` should treat as a failure.